### PR TITLE
API message searching by ANDed labels

### PR DESCRIPTION
**Pull request by rowanseymour**

### CODE REVIEW

**Comment by ericnewcomer** [comment]
What about label_required?

File: temba/api/tests.py [hunk 1]
```diff
@@ -1200,36 +1186,57 @@ def test_api_messages(self):
...
+        label1.toggle_label([msg1, msg2], add=True)
+        label2 = Label.create_unique(self.org, self.user, "Boo")
+        label2.toggle_label([msg1, msg3], add=True)
+        label3 = Label.create_unique(self.org, self.user, "Roo")
+        label3.toggle_label([msg2, msg3], add=True)
+
+        response = self.fetchJSON(url, "label=Goo&label=Boo")  # Goo or Boo
+        self.assertEqual([m['id'] for m in response.json['results']], [msg3.pk, msg2.pk, msg1.pk])
+
+        response = self.fetchJSON(url, "label=%2BGoo&label=%2BBoo")  # Goo and Boo
```
Comment (rowanseymour) [inline temba/api/tests.py]: Sucks a bit that + has to be encoded but meh

**Comment by rowanseymour** [comment]
Updated to use + and - prefixes as discussed @ericnewcomer @nicpottier

File: temba/api/tests.py (hunk 1 above)
Comment (nicpottier) [inline temba/api/tests.py]: Ya, seems like in almost all cases you would be using a library that abstracts that out for you.

File: temba/msgs/views.py [hunk 2]
```diff
@@ -836,9 +836,14 @@ def __init__(self, *args, **kwargs):
 
     def clean_name(self):
         data = self.cleaned_data['name']
+
+        if not Label.is_valid_name(data):
+            raise forms.ValidationError("Label name must not begin with + or -")
+
```
Comment (nicpottier) [inline temba/msgs/views.py]: We should add something similar for ContactGroups since we are bound to go the same way there.

**Comment by rowanseymour** [comment]
Updated so that... 
  * + and - can't occur in contact group names
  * you get a proper validation error when calling the messages action endpoint with an invalid label name

**Comment by nicpottier** [comment]
Make it so.

**Pull request merged**